src/fmm.jl

```diff
@@ -0,0 +1,296 @@
+
+    fullinteractions = []
+    compressableinteractions = []
+
+    FastBEAST.computerinteractions!(
+        test_tree,
+        trial_tree,
+        fullinteractions,
+        compressableinteractions
+    )
+
+    if threading == :single
+        for fullinteraction in fullinteractions
+            push!(
+                fullrankblocks,
+                FastBEAST.getfullmatrixview(
+                    assembler,
+                    fullinteraction[1],
+                    fullinteraction[2],
+                    Int,
+                    scalartype(operator)
+                )
+            )
+            push!(
+                correctionblocks,
+                FastBEAST.getfullmatrixview(
+                    corassembler,
+                    fullinteraction[1],
+                    fullinteraction[2],
+                    Int,
+                    scalartype(operator)
+                )
+            )
+        end
+    elseif threading == :multi
+        for i in 1:Threads.nthreads()
+            push!(fullrankblocks_perthread, MBF[])
+            push!(correctionblocks_perthread, MBF[])
+        end
+
+        Threads.@threads for fullinteraction in fullinteractions
+            push!(
+                fullrankblocks_perthread[Threads.threadid()],
+                FastBEAST.getfullmatrixview(
+                    assembler,
+                    fullinteraction[1],
+                    fullinteraction[2],
+                    Int,
+                    scalartype(operator)
+                )
+            )
+            push!(
+                correctionblocks_perthread[Threads.threadid()],
+                FastBEAST.getfullmatrixview(
+                    corassembler,
+                    fullinteraction[1],
+                    fullinteraction[2],
+                    Int,
+                    scalartype(operator)
+                )
+            )
+
+        end
+
+        for i in eachindex(fullrankblocks_perthread)
+            append!(fullrankblocks, fullrankblocks_perthread[i])
+        end
+        for i in eachindex(correctionblocks_perthread)
+            append!(correctionblocks, correctionblocks_perthread[i])
+        end
+    end
+
+    return fullrankblocks, correctionblocks, fullinteractions
+end
+
+# mapping mesh to quadrature points
+function meshtopoints(X::BEAST.Space, quadorder)
+
+    test_elements, _, _ = assemblydata(X)
+    tshapes = refspace(X)
+
+    test_eval(x) = tshapes(x)
+    qp = BEAST.quadpoints(test_eval,  test_elements, (quadorder,))
+    points = zeros(Float64, length(qp) * length(qp[1,1]), 3)
+    ind = 1
+    for el in qp
+        for i in el
+            points[ind, 1] = i.point.cart[1]
+            points[ind, 2] = i.point.cart[2]
+            points[ind, 3] = i.point.cart[3]
+            ind += 1
+        end
+    end
+    
+    return points, qp
+end
+
+# construction of B matrix, if Ax = b and A = transpose(B)GB
```

Comment:
I am thinking that maybe should remaining the functions to something more akin to

sample_basis()
and
sample_curlofbasis()